ocamlPackages.smtml: init at 0.7.0

ethancedwards8 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done
OCaml dependies for the `owi` package: https://github.com/NixOS/nixpkgs/pull/414133
The `owi` package only builds on ocaml version 5.3, not 5.1. However, I have to use 5.1 because the `stdcompat` library does not build on 5.3 and I am waiting on upstream to support it. It is in the works. I thought I would add the package deps so that they are in nixpkgs and available for as many people as possible. 

This work is in preparation for Summer of Nix 2025, funded by the NLNet and NixOS Foundation, among others.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ethancedwards8 committed:
ocamlPackages.hc: init at 0.5

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
ocamlPackages.prelude: init at 0.5

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
ocamlPackages.scfg: init at 0.5

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
ocamlPackages.patricia-tree: init at 0.11.0

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
ocamlPackages.smtml: init at 0.7.0

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 commented:
Not sure why these didn't trigger any new builds?..

vbgl commented on pkgs/development/ocaml-modules/hc/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  # upstream git server is misconfigured and cannot be cloned
+  src = fetchurl {
+    url = "https://git.zapashcanon.fr/zapashcanon/hc/archive/${version}.tar.gz";
```
Is it safe to use `fetchurl` with this URL? `fetchzip` might be more adequate.

ethancedwards8 replied:
I'm not familiar with the safety of `fetchurl` do you have anything for me to read?

vbgl commented on pkgs/development/ocaml-modules/patricia-tree/default.nix:
```diff
@@ -0,0 +1,61 @@
+    hash = "sha256-lpmU0KhsyIHxPBiw38ssA7XFEMsRvOT03MByoJG88Xs=";
+  };
+
+  strictDeps = true;
+
```
```suggestion
```

This line is redundant.

ethancedwards8 replied:
Why? I have it in all the other new packages.

vbgl commented on pkgs/development/ocaml-modules/patricia-tree/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    mdx
+  ];
```
```suggestion
  nativeCheckInputs = [
    mdx.mdx
  ];
```

vbgl replied:
Sorry, there was a typo in my previous comment. The differences are twofold: `mdx` is a check-input (so not kept as a dependency in case you set `doCheck` to false); and only the binary is needed on the “native” platform.

vbgl commented on pkgs/development/ocaml-modules/patricia-tree/default.nix:
```diff
@@ -0,0 +1,61 @@
+    mdx
+  ];
+
+  buildInputs = [
+    findlib
+    astring
+    mdx
+    fmt
+    camlp-streams
+    csexp
+    ocaml-version
+    qcheck-core
+    re
+  ];
+
+  checkInputs = [
+    ppx_inline_test
+  ];
```
```suggestion
  checkInputs = [
    mdx
    ppx_inline_test
    qcheck-core
  ];
```

vbgl replied:
Are you sure?

vbgl commented on pkgs/development/ocaml-modules/patricia-tree/default.nix:
```diff
@@ -0,0 +1,61 @@
+{
+  lib,
+  buildDunePackage,
+  fetchFromGitHub,
+  findlib,
+  mdx,
+  qcheck-core,
+  re,
+  camlp-streams,
+  csexp,
+  astring,
+  ocaml-version,
+  fmt,
+  logs,
+  ppx_inline_test,
```
```suggestion
  lib,
  buildDunePackage,
  fetchFromGitHub,
  mdx,
  qcheck-core,
  ppx_inline_test,
```

ethancedwards8 replied:
The package does not build without these? I do not understand.

ethancedwards8 committed:
ocamlPackages.processor: init at 0.1

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 414281`

Logs: https://github.com/ethancedwards8/nixpkgs-review-gha/actions/runs/15503904714

ethancedwards8 committed:
ocamlPackages.pyml: fix tests in sandbox

This package is disabled on the latest ocaml version, so its test
failures went under the radar. I discovered it by trying to build it
with ocaml-ng.ocamlPackages_5_1.pyml.

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

vbgl commented on pkgs/development/ocaml-modules/scfg/default.nix:
```diff
@@ -0,0 +1,57 @@
+  ];
+
+  checkInputs = [
+    writableTmpDirAsHomeHook
```
This is a cool workaround, but really there is an upstream bug (they are not testing what they intend to test) that you may want to report.

vbgl reviewed (commented):
Please also double-check the `smtml` derivation: there are some issue there that you should be able to fix in light of the above comments.

ethancedwards8 commented:
These are all a dependency of a package I am working on, so they need to be included in one PR fwiw.

vbgl commented:
No, they don’t “need” to. For instance, the first commit in your other PR, that adds `hc` looks ready and could be merged. But now it is stuck by the other commits that need polishing or discussion. The problem is that any review of that PR will need to look again (and again) at this commit (e.g. just to be sure it did not get broken by accident during a rebase), creating unnecessary work.

ethancedwards8 commented on pkgs/development/ocaml-modules/scfg/default.nix:
```diff
@@ -0,0 +1,57 @@
+    menhir
+  ];
+
+  buildInputs = [
+    bos
+    cmdliner
+    fmt
+    fpath
+    menhir
+    menhirLib
+    prelude
+    sedlex
+  ];
```
I don't believe this is correct. The package fails to build with this.
